data/quodlibet.1
.\" Man page generated from reStructuredText.
.
.TH QUODLIBET 1 "" "" ""
.SH NAME
quodlibet \- audio library manager and player
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.nf
\fBquodlibet\fP [ \fB\-\-print\-playing\fP | \fIcontrol\fP ]
\fBexfalso\fP [ \fIdirectory\fP ]
.fi
.sp
.SH DESCRIPTION
.sp
Quod Libet is a music management program. It provides several different
ways to view your audio library, as well as support for Internet radio and
audio feeds. It has extremely flexible metadata tag editing and searching
capabilities.
.sp
This manual page is only a short reference for Quod Libet. Complete
documentation is available at
\fI\%https://quodlibet.readthedocs.io/en/latest/guide/index.html\fP\&.
.SH OPTIONS
.INDENT 0.0
.TP
.BI \-\-add\-location\fB= <location>
Add a file or directory to the library
.TP
.B \-\-debug
Print debugging information
.TP
.BI \-\-enqueue\fB= <filename|query>
Enqueue a filename or query results
.TP
.BI \-\-enqueue\-files\fB= <filename[,filename..]>
Enqueue comma\-separated files
.TP
.BI \-\-filter \ <tag=value>
Filter on a tag value
.TP
.B \-\-focus
Focus the running player
.TP
.B \-\-force\-previous
Jump to previous song
.TP
.B \-\-help
Display brief usage information
.TP
.B \-\-hide\-window
Hide main window
.TP
.B \-\-list\-browsers
List available browsers
.TP
.B \-\-next
Jump to next song
.TP
.B \-\-no\-plugins
Start without plugins
.TP
.BI \-\-open\-browser\fB= BrowserName
Open a new browser
.TP
.B \-\-pause
Pause playback
.TP
.B \-\-play
Start playback
.TP
.BI \-\-play\-file\fB= filename
Play a file
.TP
.B \-\-play\-pause
Toggle play/pause mode
.TP
.B \-\-previous
Jump to previous song if near the beginning, otherwise restart
.TP
.B \-\-print\-playing
Print out information about the currently playing song. You may
provide in a string like the kind described in the RENAMING FILES
section below.
.TP
.B \-\-print\-playlist
Print the current playlist
.TP
.BI \-\-print\-query\fB= <query>
Print filenames of results of query
.TP
.B \-\-print\-query\-text
Print the active text query
.TP
.B \-\-print\-queue
Print the contents of the queue
.TP
.BI \-\-query\fB= search\-string
Search your audio library
.TP
.BI \-\-queue\fB= <on|off|t>
Show or hide the queue
.TP
.B \-\-quit
Exit Quod Libet
.TP
.BI \-\-random\fB= tag
Filter on a random value
.TP
.BI \-\-rating\fB= <[+|\-]0.0..1.0>
Set rating of playing song
.TP
.B \-\-rating\-down
Decrease rating of playing song by one star
.TP
.B \-\-rating\-up
Increase rating of playing song by one star
.TP
.B \-\-refresh
Refresh and rescan library
.TP
.BI \-\-repeat\fB= <off|on|t>
Turn repeat off, on, or toggle
.TP
.BI \-\-repeat\-type\fB= <current|all|one|off>
Repeat the currently playing song, the current list, stop after
one song, or turn repeat off
.TP
.B \-\-run
Start Quod Libet if it isn\(aqt running
.TP
.BI \-\-seek\fB= <[+|\-][HH:]MM:SS>
Seek within the playing song
.TP
.BI \-\-set\-browser\fB= BrowserName
Set the current browser
.TP
.B \-\-show\-window
Show main window
.TP
.BI \-\-shuffle\fB= <off|on|t>
Turn shuffle off, on, or toggle
.TP
.BI \-\-shuffle\-type\fB= <random|weighted|off>
Set the shuffle type to be random, to prefer higher rated songs,
or turn shuffle off
.TP
.B \-\-start\-hidden
Don\(aqt show any windows on start
.TP
.B \-\-start\-playing
Begin playing immediately
.TP
.B \-\-status
Print playing status
.TP
.B \-\-stop
Stop playback
.TP
.BI \-\-stop\-after\fB= <0|1|t>
Stop after the playing song
.TP
.B \-\-toggle\-window
Toggle main window visibility
.TP
.B \-\-unfilter
Remove active browser filters
.TP
.BI \-\-unqueue\fB= <filename|query>
Unqueue a file or query
.TP
.B \-\-version
Display version and copyright
.TP
.BI \-\-volume\fB= <[+|\-]0..100>
Set the volume
.TP
.B \-\-volume\-down
Turn down the volume
.TP
.B \-\-volume\-up
Turn up the volume
.TP
.BI \-\-with\-pattern\fB= <pattern>
Set the output format of the --print-playing, --print-playlist,
--print-query, and --print-queue commands.  You may provide in a
tag-pattern string like the kind described in the RENAMING FILES
section below.
.UNINDENT
.SH ALBUM COVERS
.sp
Album covers should be put in the same directory as the songs they apply
to, and have "folder", "front", or "cover" in their filenames. If you want
to store multiple albums in the same directory but keep distinct cover
images, the name of the appropriate image file must contain the labelid tag
value, e.g. COCX\-32760 cover.jpg.
.SH TIED TAGS
.sp
Many places in Quod Libet allow you to use "tied tags". Tied tags are two
tag names joined together with a "~" like "title~version" or "album~part".
Tied tags result in "nice" displays even when one of the tags is missing;
for example, "title~version" will result in Title \- Version when a version
tag is present, but only Title when one isn\(aqt. You can tie any number of
tags together.
.SH SEARCH SYNTAX
.sp
All of Quod  Libet\(aqs search boxes support advanced searches of the
following forms:
.sp

.nf
tag = value
tag = !value
tag = "value"
tag = /value/
tag = &(value1, value2)
tag = |(value1, value2)
!tag = value
|(tag1 = value1, tag2 = value2)
&(tag1 = value1, tag2 = value2)
#(numerictag < value)
#(numerictag = value)
#(numerictag > value)
.fi
.sp
.sp

.sp
The \(aqc\(aq postfix on strings or regular expressions makes the  search
case\-sensitive. Numeric values may be given as integers, floating\-point
numbers, MM:SS format, or simple English, e.g. "3 days", "2 hours".
.sp
See \fI\%https://quodlibet.readthedocs.io/en/latest/guide/searching.html\fP\&.
.sp
All internal tags begin with a ~ character. Non\-numeric internal tags are
~basename, ~dirname, ~filename, ~format, ~length, ~people, and ~rating.
Numeric internal tags are ~#added, ~#bitrate, ~#disc, ~#lastplayed,
~#laststarted, ~#length, ~#mtime, ~#playcount, ~#skipcount, and ~#track.
.sp
See \fI\%https://quodlibet.readthedocs.io/en/latest/guide/tags/internal_tags.html\fP\&.
.SH RENAMING FILES
.sp
Quod Libet allows you to rename files based on their tags. In some cases
you may wish to alter the filename depending on whether some tags are
present or missing, in addition to their values. A common pattern might be
.sp
\fB<tracknumber>. <title~version>\fP
.sp
You can use a \(aq|\(aq to only text when a tag is present:
.sp
\fB<tracknumber|<tracknumber>. ><title~version>\fP
.sp
You can also specify literal text to use if the tag is missing by adding another \(aq|\(aq:
.sp
\fB<album|<album>|No Album> \- <title>\fP
.sp
See \fI\%https://quodlibet.readthedocs.io/en/latest/guide/renaming_files.html\fP\&.
.SH AUDIO BACKENDS
.sp
Quod Libet uses GStreamer for audio playback. It tries to read your GConf
GStreamer configuration, but if that fails it falls back to osssink. You can
change the pipeline option in ~/.quodlibet/config to use a different sink, or
pass options to the sink. For example, you might use esdsink or alsasink
device=hw:1.
.sp
See \fI\%https://quodlibet.readthedocs.io/en/latest/guide/playback/backends.html\fP\&.
.SH FILES
.INDENT 0.0
.TP
.B ~/.quodlibet/songs
A pickled Python dict of cached metadata. Deleting this file will remove all
songs from your library.
.TP
.B ~/.quodlibet/config
Quod Libet\(aqs configuration file. This file is overwritten when Quod Libet
exits.
.TP
.B ~/.quodlibet/current
A "key=value" file containing information about the currently playing song.
.TP
.B ~/.quodlibet/control
A FIFO connected to the most\-recently\-started instance of the program.
\-\-next, \-\-previous, etc., use this to control the player.
.TP
.B ~/.quodlibet/plugins/
Put plugins here.
.TP
.B ~/.quodlibet/browsers/
Put custom library browsers here.
.UNINDENT
.sp
See \fI\%https://quodlibet.readthedocs.io/en/latest/guide/interacting.html\fP\&.
.SH BUGS
.sp
See \fI\%https://github.com/quodlibet/quodlibet/issues\fP for a list of all
currently open bugs and feature requests.
.SH AUTHORS
.sp
Joe Wreschnig and Michael Urman are the primary authors of Quod Libet.
.SH SEE ALSO
.nf
\fI\%https://quodlibet.readthedocs.io/en/latest/guide/\fP,
\fI\%https://quodlibet.readthedocs.io/en/latest/guide/faq.html\fP,
\fBregex\fP(7), \fBgst\-launch\fP(1)
.fi
.sp
.\" Generated by docutils manpage writer.
.


NEWS.rst
.. _News:

.. _release-4.7.1:

4.7.1 - T-rex in: popular corn
------------------------------

ℹ️ A highlight of the changes in 4.7.1 is below.
For full details,
see `the 4.7 commit log <https://github.com/quodlibet/quodlibet/compare/release-4.6.0...release-4.7.1>`_.

Packaging Changes
 * Quodlibet now requires Python 3.10+ :pr:`4400` (:user:`Nick Boultbee <declension>`),
   and supports Python 3.12.
 * Quodlibet now uses ``gtksourceview4`` optionally instead of ``gtksourceview3`` :pr:`4377`
 * Appdata is now installed into ``/usr/share/metainfo`` instead of ``/usr/share/appdata`` :pr:`4375`
 * The tests now use use ``appstream`` optionally instead of ``appstream-util`` :pr:`4376`
 * Move to zlib 1.3.1.  The previous version is no longer available upstream. :pr:`4491` (:user:`John Dykstra`)
 * Update OS X build instructions re Python 3.12 :pr:`4446` (:user:`John Dykstra`)
 * libsoup 3 also needs ``gir1.2-soup-3.0`` :pr:`4391` (:user:`lbt`)
 * Update feedparser to avoid deprecated cgi :pr:`4479` (:user:`Nick Boultbee <declension>`)

General
 * Advanced prefs: add option to use playbin3 instead of playbin :pr:`4693` (:user:`Christoph Reiter <lazka>`)
 * Auto-enable more plugins: :pr:`4667` (:user:`Nick Boultbee <declension>`)
 * Tidy EF prefs :pr:`4627` (:user:`Nick Boultbee <declension>`)
 * Exfalso: add progress bar (fixes #4604) :pr:`4625` (:user:`abretaud`)
 * Waveform seekbar: update config :pr:`4609` (:user:`Nick Boultbee <declension>`)
 * Fix logging time, and quieten logs a bit :pr:`4611` (:user:`Nick Boultbee <declension>`)
 * Squeezebox plugins: Migrate to raw socket connection :pr:`4563` (:user:`luk1337`)
 * Document how to run exfalso and quodlibet from source :pr:`4569` (:user:`sezanzeb`)
 * Fix: Dereference symbolic links when writing to file (Issue 4543) :pr:`4548` (:user:`Peptis`)
 * Make waveform height configurable (in plugin configuration) :pr:`4550` (:user:`eudoxos`)
 * Sort list of active plugins in config file :pr:`4444` (:user:`toy`)
 * Fix shuffle history (2) :pr:`4405` (:user:`slosd`)
 * Paned Browser: small refactors / tidyings :pr:`4421` (:user:`Nick Boultbee <declension>`)
 * Fix paned browser total rendering (#4415) :pr:`4420` (:user:`Nick Boultbee <declension>`)
 * Update user guide for QL version of Fallback tags :pr:`4418` (:user:`Nick Boultbee <declension>`)
 * operon: multiple file edit :pr:`3418` (:user:`remvee`)
 * Advanced prefs pane boolean switches :pr:`4411` (:user:`jej`)
 * Merge advanced prefs plugin into prefs pane :pr:`4408` (:user:`jej`)
 * Revert "Fix shuffle history (#4255)" :pr:`4382` (:user:`Nick Boultbee <declension>`)
 * Add disjunction / fallback tag expression :pr:`4372` (:user:`afontenot`)
 * Repeat should pass set_explicit and set_implicit instead of set :pr:`4341` (:user:`toy`)
 * Connect TreeViewHints->view leave-notify-event to __undisplay :pr:`4365` (:user:`luk1337`)
 * Explicitly stop playing song before deleting it :pr:`4689` (:user:`Nick Boultbee <declension>`)
 * Add genre splitting in tag editor :pr:`4250` (:user:`scresante`)
 * Make Sort Tags Plugin: restrict and improve behaviour :pr:`4633` (:user:`Nick Boultbee <declension>`)
 * Create Sort Tags plugin: never create empty tags :pr:`4630` (:user:`Nick Boultbee <declension>`)
 * Fix playlist renaming (#4589) :pr:`4616` (:user:`Nick Boultbee <declension>`)
 * Nicer DBUS plugin text / explanation :pr:`4615` (:user:`Nick Boultbee <declension>`)
 * Fix default plugins (#4612) :pr:`4614` (:user:`Nick Boultbee <declension>`)
 * Parse year from MM/DD/YYYY dates :pr:`4556` (:user:`Iris Ilexiris`)
 * TrackNumbers: Up limit to 9999 :pr:`4471` (:user:`Adam Chalkley`)
 * PlaycountEqualizer extension improvements :pr:`4499` (:user:`Rumato Estorsky`)
 * Fix / improve Gnome SearchProvider :pr:`4501` (:user:`Nick Boultbee <declension>`)
 * Add option to hide timestamps from lyrics panel :pr:`4523` (:user:`Vic Verstraete`)
 * Include configured title/album details in MPRIS plugin :pr:`4493` (:user:`Markus Koller`)
 * Only scrobble over HTTPS :pr:`4487` (:user:`Nick Boultbee <declension>`)
 * Squeezebox plugins: fail nicely if telnetlib is missing :pr:`4480` (:user:`Nick Boultbee <declension>`)
 * Docs: Add unofficial macOS 4.6.0 builds :pr:`4467` (:user:`Nick Boultbee <declension>`)
 * Different approach to unsortable columns :pr:`4468` (:user:`Nick Boultbee <declension>`)
 * Pass the name of the playlist to create (#4392) :pr:`4430` (:user:`David Greaves`)
 * Fix the clickability when always allowing sorting :pr:`4453` (:user:`Nick Boultbee <declension>`)
 * Support producer tag properly :pr:`4452` (:user:`Nick Boultbee <declension>`)
 * Unify playlist hidden file check (so be Windows-aware) :pr:`4690` (:user:`Nick Boultbee <declension>`)

Developer
 * Upgrade all Python syntax to 3.10 :pr:`4407` (:user:`Nick Boultbee <declension>`)
 * Migrate lots of old unittest assertions to native :pr:`4617` (:user:`Nick Boultbee <declension>`)
 * Remove redundant GTK checks / hacks :pr:`4412` (:user:`Nick Boultbee <declension>`)
 * Add Nix flake for dev :pr:`4401` (:user:`Nick Boultbee <declension>`)
 * Expand ruff analysis rules, fix things, and autoformat :pr:`4686` (:user:`Nick Boultbee <declension>`)
 * Update linting & formatting throughout codebase :pr:`4634` (:user:`Nick Boultbee <declension>`)
 * CI: switch Fedora to 39 :pr:`4481` (:user:`Nick Boultbee <declension>`)
 * Require poetry v2 and update pyproject.toml :pr:`4670` (:user:`Christoph Reiter <lazka>`)
 * Update MacOS build to harfbuzz 10.1.0. :pr:`4639` (:user:`jdykstra`)
 * tests: Filter out unsupported file formats properly :pr:`4566` (:user:`luk1337`)
 * Repo code quality upgrade (with Ruff) :pr:`4404` (:user:`Nick Boultbee <declension>`)
 * Move to Ruff for linting :pr:`4403` (:user:`Nick Boultbee <declension>`)
 * gdist: Fix filtering dirs logic in get_missing() :pr:`4577` (:user:`luk1337`)
 * Clean up loose ends in OS X support :pr:`4443` (:user:`John Dykstra`)

data/quodlibet.bash
# quodlibet(1) completion                                  -*- shell-script -*-
# vim: sts=4 sw=4 et
# shellcheck disable=SC2207
#
# This file is part of Quodlibet.
#
# Copyright 2019 Arnaud Rebillout
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

# Quodlibet tags

_ql_tags_make() {
    command -v operon >/dev/null 2>&1 || return
    operon tags -a -t -c tag | tr '\n' ' '
}

_ql_tags="$(_ql_tags_make)"

# Quodlibet supported extensions (as bash globs)

_ql_audio_glob_make() {
    command -v python3 >/dev/null 2>&1 || return
    command -v quodlibet >/dev/null 2>&1 || return

    python3 - <<EOF

init()
exts = []
for k in sorted(loaders.keys()):
    dot, ext = k.split('.', 1)
    exts.append(ext)
print('@(' + '|'.join(exts) + ')')
EOF
}

_ql_audio_glob="$(_ql_audio_glob_make)"
_ql_img_glob="@(gif|jp?(e)g|png)"

# Quodlibet completion

_quodlibet() {
    # Assigned variable by _init_completion:
    #   cur    Current argument.
    #   prev   Previous argument.
    #   words  Argument array.
    #   cword  Argument array size.
    local cur prev words cword
    _init_completion -n = || return

    # Quodlibet commands and options
    local opts=(--add-location --debug --enqueue --enqueue-files
        --filter --focus --force-previous --help --hide-window
        --list-browsers --next --no-plugins --open-browser --pause
        --play --play-file --play-pause --previous --print-playing
        --print-playlist --print-query --print-query-text --print-queue
        --query --queue --quit --random --refresh --repeat --repeat-type
        --run --seek --set-browser --set-rating --show-window --shuffle
        --shuffle-type --start-hidden --start-playing --status --stop
        --stop-after --toggle-window --unfilter --unqueue --version
        --volume --volume-down --volume-up)

    # Completion per otion
    case "$prev" in

    --enqueue | --unqueue | --enqueue-files | --add-location | --play-file)
        # For these options we complete with audio files, even though
        # it's not necessarily what user wants.
        # --add-location=location
        # --enqueue=filename|query
        # --enqueue-files=filename[,filename..]
        # --play-file=filename
        # --unqueue=filename|query
        _filedir "$_ql_audio_glob"
        return 0
        ;;

    --queue | --repeat | --shuffle)
        compopt -o nosort
        comps="on off toggle"
        COMPREPLY=($(compgen -W "$comps" -- "$cur"))
        return 0
        ;;

    --repeat-type)
        compopt -o nosort
        comps="current all one off"
        COMPREPLY=($(compgen -W "$comps" -- "$cur"))
        return 0
        ;;

    --shuffle-type)
        compopt -o nosort
        comps="random weighted off"
        COMPREPLY=($(compgen -W "$comps" -- "$cur"))
        return 0
        ;;

    --stop-after)
        compopt -o nosort
        comps="0 1 t"
        COMPREPLY=($(compgen -W "$comps" -- "$cur"))
        return 0
        ;;

    --filter) # tag=value
        compopt -o nospace
        comps="$_ql_tags"
        COMPREPLY=($(compgen -S= -W "$comps" -- "$cur"))
        return 0
        ;;

    --random) # tag
        comps="$_ql_tags"
        COMPREPLY=($(compgen -W "$comps" -- "$cur"))
        return 0
        ;;

    --open-browser | --query | --print-query | --seek | --set-browser | --set-rating | --volume | --with-pattern)
        # These options expect an argument that we can't guess.
        # --open-browser=BrowserName
        # --query=query
        # --print-query=query
        # --seek=[+|-][HH:]MM:SS
        # --set-browser=BrowserName
        # --set-rating=0.0..1.0
        # --volume=(+|-|)0..100
        # --with-pattern=pattern
        return 0
        ;;

    -*)
        # Other options don't expect any argument, however user might want
        # to provide more options. So we don't return yet.
        ;;

    esac

    # Initial completion
    case "$cur" in
    -*)
        COMPREPLY=($(compgen -W "${opts[*]}" -- "$cur"))
        return 0
        ;;
    *)
        return 0
        ;;
    esac
} &&
    complete -F _quodlibet quodlibet

# Operon completion

_in_array() {
    local i
    for i in "${@:2}"; do
        [[ $1 == "$i" ]] && return
    done
}

_operon() {
    # Assigned variable by _init_completion:
    #   cur    Current argument.
    #   prev   Previous argument.
    #   words  Argument array.
    #   cword  Argument array size.
    # shellcheck disable=SC2034
    local cur prev words cword
    _init_completion -n = || return

    # Operon commands and options
    local opts=(--help --verbose --version)
    local cmds=(add clear copy edit fill help
        image-clear image-extract image-set
        info list print remove set tags)

    # Check if a command was entered already
    local command i
    for ((i = 0; i < ${#words[@]} - 1; i++)); do
        if _in_array "${words[i]}" "${cmds[@]}"; then
            command=${words[i]}
            break
        fi
    done

    # Completion per command
    if [[ -n $command ]]; then
        case $command in
        add) # <tag> <value> <file> [<files>]
            nargs=0
            for ((j = i + 1; j < ${#words[@]} - 1; j++)); do
                case "${words[j]}" in
                -*) continue ;;
                *) nargs=$((nargs + 1)) ;;
                esac
            done
            if [[ $nargs -eq 0 ]]; then
                comps="$_ql_tags"
                COMPREPLY=($(compgen -W "$comps" -- "$cur"))
            elif [[ $nargs -ge 2 ]]; then
                _filedir "$_ql_audio_glob"
            fi
            return 0
            ;;

        clear) # [--dry-run] [-a | -e <pattern> | <tag>] <file> [<files>]
            case $cur in
            -*)
                comps="--all --dry-run --regexp"
                COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                ;;
            *)
                case $prev in
                -a | --all)
                    compopt -o default
                    COMPREPLY=()
                    ;;
                -e | --regex) ;;
                *)
                    nargs=0
                    for ((j = i + 1; j < ${#words[@]} - 1; j++)); do
                        case "${words[j]}" in
                        -d | --dry-run) continue ;;
                        -e | --regexp) continue ;;
                        -a | --all) nargs=$((nargs + 1)) ;;
                        *) nargs=$((nargs + 1)) ;;
                        esac
                    done
                    if [[ $nargs -eq 0 ]]; then
                        comps="$_ql_tags"
                        COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                    elif [[ $nargs -ge 1 ]]; then
                        _filedir "$_ql_audio_glob"
                    fi
                    ;;
                esac
                ;;
            esac
            return 0
            ;;

        copy) # [--dry-run] [--ignore-errors] <source> <dest>
            case $cur in
            -*)
                comps="--dry-run --ignore-errors"
                COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                ;;
            *)
                _filedir "$_ql_audio_glob"
                ;;
            esac
            return 0
            ;;

        edit) # [--dry-run] <file>
            case $cur in
            -*)
                comps="--dry-run"
                COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                ;;
            *)
                _filedir "$_ql_audio_glob"
                ;;
            esac
            return 0
            ;;

        fill) # [--dry-run] <pattern> <file> [<files>]
            case $cur in
            -*)
                comps="--dry-run"
                COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                ;;
            *)
                # no completion for <pattern>
                nargs=0
                for ((j = i + 1; j < ${#words[@]} - 1; j++)); do
                    case "${words[j]}" in
                    -*) continue ;;
                    *) nargs=$((nargs + 1)) ;;
                    esac
                done
                if [[ $nargs -ge 1 ]]; then
                    _filedir "$_ql_audio_glob"
                fi
                ;;
            esac
            return 0
            ;;

        help) # [<command>]
            COMPREPLY=($(compgen -W "${cmds[*]}" -- "$cur"))
            return 0
            ;;

        image-clear) # <file> [<files>]
            _filedir "$_ql_audio_glob"
            return 0
            ;;

        image-extract) # [--dry-run] [--primary] [-d <destination>] <file> [<files>]
            case $cur in
            -*)
                comps="--destination --dry-run --primary"
                COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                ;;
            *)
                case $prev in
                -d | --destination)
                    compopt -o default
                    COMPREPLY=()
                    ;;
                *)
                    _filedir "$_ql_audio_glob"
                    ;;
                esac
                ;;
            esac
            return 0
            ;;

        image-set) # <image-file> <file> [<files>]
            case $prev in
            image-set)
                _filedir "$_ql_img_glob"
                ;;
            *)
                _filedir "$_ql_audio_glob"
                ;;
            esac
            return 0
            ;;

        info) # [-t] [-c <c1>,<c2>...] <file>
            case $cur in
            -*)
                comps="--columns --terse"
                COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                ;;
            *)
                case $prev in
                -c | --columns) ;;
                *)
                    _filedir "$_ql_audio_glob"
                    ;;
                esac
                ;;
            esac
            return 0
            ;;

        list) # [-a] [-t] [-c <c1>,<c2>...] <file>
            case $cur in
            -*)
                comps="--all --columns --terse"
                COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                ;;
            *)
                case $prev in
                -c | --columns) ;;
                *)
                    _filedir "$_ql_audio_glob"
                    ;;
                esac
                ;;
            esac
            return 0
            ;;

        print) # [-p <pattern>] <file> [<files>]
            case $cur in
            -*)
                comps="--pattern"
                COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                ;;
            *)
                case $prev in
                -p | --pattern) ;;
                *)
                    _filedir "$_ql_audio_glob"
                    ;;
                esac
                ;;
            esac
            return 0
            ;;

        remove) # [--dry-run] <tag> [-e <pattern> | <value>] <file> [<files>]
            case $cur in
            -*)
                comps="--dry-run --regexp"
                COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                ;;
            *)
                nargs=0
                for ((j = i + 1; j < ${#words[@]} - 1; j++)); do
                    case "${words[j]}" in
                    -*) continue ;;
                    *) nargs=$((nargs + 1)) ;;
                    esac
                done
                if [[ $nargs -eq 0 ]]; then
                    comps="$_ql_tags"
                    COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                elif [[ $nargs -ge 2 ]]; then
                    _filedir "$_ql_audio_glob"
                fi
                ;;
            esac
            return 0
            ;;

        set) # [--dry-run] <tag> <value> <file> [<files>]
            case $cur in
            -*)
                comps="--dry-run"
                COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                ;;
            *)
                nargs=0
                for ((j = i + 1; j < ${#words[@]} - 1; j++)); do
                    case "${words[j]}" in
                    -*) continue ;;
                    *) nargs=$((nargs + 1)) ;;
                    esac
                done
                if [[ $nargs -eq 0 ]]; then
                    comps="$_ql_tags"
                    COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                elif [[ $nargs -ge 2 ]]; then
                    _filedir "$_ql_audio_glob"
                fi
                ;;
            esac
            return 0
            ;;

        tags) # [-t] [-c <c1>,<c2>...]
            case $cur in
            -*)
                comps="--all --columns --terse"
                COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                ;;
            esac
            return 0
            ;;
        esac
    fi

    # Initial completion
    case "$cur" in
    -*)
        COMPREPLY=($(compgen -W "${opts[*]}" -- "$cur"))
        return 0
        ;;
    *)
        COMPREPLY=($(compgen -W "${cmds[*]}" -- "$cur"))
        return 0
        ;;
    esac

} &&
    complete -F _operon operon
